manim/mobject/functions.py

```diff
@@ -147,3 +152,67 @@ def get_function(self):
 
     def get_point_from_function(self, x):
         return self.parametric_function(x)
+
+
+class ImplicitFunction(VMobject, metaclass=ConvertToOpenGL):
+    def __init__(
+        self,
+        func: Callable[[float, float], float],
+        x_range: Optional[Sequence[float]] = None,
+        y_range: Optional[Sequence[float]] = None,
+        min_depth: int = 5,
+        max_quads: int = 2500,
+        **kwargs
+    ):
+        """An implicit function.
+
+        Parameters
+        ----------
+        func
+            The implicit function in the form ``f(x, y) = 0``.
+        x_range
+            The x min and max of the function.
+        y_range
+            The y min and max of the function.
+        min_depth
+            The minimum depth of the function to calculate.
+        max_quads
+            The maximum number of quadtrees to use.
+
+        Examples
+        --------
+        .. manim:: ImplicitFunctionExample
+            :save_last_frame:
+
+            class ImplicitFunctionExample(Scene):
+                def construct(self):
+                    graph = ImplicitFunction(
+                        lambda x, y: x * y ** 2 - x ** 2 * y - 2,
+                        color=YELLOW
+                    )
+                    self.add(NumberPlane(), graph)
+        """
+        x_range = x_range or [
+            -config.frame_width / 2,
+            config.frame_width / 2,
+        ]
+        y_range = y_range or [
+            -config.frame_height / 2,
+            config.frame_height / 2,
+        ]
+        p_min, p_max = (
+            np.array([x_range[0], y_range[0]]),
+            np.array([x_range[1], y_range[1]]),
+        )
+
+        points = plot_isoline(
+            lambda u: func(u[0], u[1]),
+            p_min,
+            p_max,
+            min_depth,
+            max_quads,
+        )
+        points = [np.pad(i, [(0, 0), (0, 1)]) for i in points]
+        super().__init__()
+        for p in points:
+            self.add(Line(**kwargs).set_points_as_corners(p))
```

Comment:
I agree with replicating generate points method as ParametricFunction has done.

> In addition, the smoothing method also used in ParametricFunction can also be used here

I also think we should make smoothing available, but not by default, because quite a lot of implicit graphs are not smooth and do twist a lot, which further explains why a lot of points are required. @jared-hughes mentioned how they could improve that specifically by implementing a linearity test. But that's mostly the internals, which could be updated later.

Reply:
Can you provide an example where smoothing would not be appropriate? It seems like it would work fine for all the examples you've provided. I agree that it should be a parameter that can be changed by the user.

Reply:
Something that comes to mind is the discontinuous functions like `np.sin(1 / x)` which currently doesn't look all nice, but that might just be that not enough points are being sampled.
![Test_ManimCE_v0 10 0](https://user-images.githubusercontent.com/83535735/135032469-9320fd6b-4955-4a88-8050-f3037d06ff2d.png)

But in multitudes of general cases, I think it's fine to default apply smoothing, if that's preferred.

Reply:
Yea, I think we should apply smoothing by default. Can you then maybe try lowering the default quads to see what still gives a decent plot for most implicit functions (not pathological ones like sin(1/x))?